File: solutions/380-insert-delete-getrandom-o1/solution.py
import random


class RandomizedSet:
    def __init__(self):
        self.val_list = []
        self.val_map = {}

    def insert(self, val: int) -> bool:
        if val in self.val_map:
            return False
        self.val_map[val] = len(self.val_list)
        self.val_list.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.val_map:
            return False
        idx = self.val_map[val]
        last = self.val_list[-1]
        self.val_list[idx] = last
        self.val_map[last] = idx
        self.val_list.pop()
        del self.val_map[val]
        return True

    def get_random(self) -> int:
        return random.choice(self.val_list)


if __name__ == "__main__":
    s = RandomizedSet()
    r1 = s.insert(1)
    r2 = s.insert(2)
    r3 = s.insert(1)
    r4 = s.remove(2)
    r5 = s.insert(2)
    ok = r1 and r2 and not r3 and r4 and r5
    status = "PASS" if ok else "FAIL"
    print(f"{status} | Insert(1)={r1} Insert(2)={r2} Insert(1)={r3} Remove(2)={r4} Insert(2)={r5}")

File: solutions/380-insert-delete-getrandom-o1/solution.go
package main

import (
	"fmt"
	"math/rand"
)

type RandomizedSet struct {
	valList []int
	valMap  map[int]int
}

func NewRandomizedSet() RandomizedSet {
	valList := make([]int, 0)
	valMap := make(map[int]int)
	return RandomizedSet{valList: valList, valMap: valMap}
}

func (this *RandomizedSet) Insert(val int) bool {
	if _, ok := this.valMap[val]; ok {
		return false
	}
	this.valMap[val] = len(this.valList)
	this.valList = append(this.valList, val)
	return true
}

func (this *RandomizedSet) Remove(val int) bool {
	idx, ok := this.valMap[val]
	if !ok {
		return false
	}
	last := this.valList[len(this.valList)-1]
	this.valList[idx] = last
	this.valMap[last] = idx
	this.valList = this.valList[:len(this.valList)-1]
	delete(this.valMap, val)
	return true
}

func (this *RandomizedSet) GetRandom() int {
	return this.valList[rand.Intn(len(this.valList))]
}

func main() {
	s := NewRandomizedSet()
	r1 := s.Insert(1)
	r2 := s.Insert(2)
	r3 := s.Insert(1)
	r4 := s.Remove(2)
	r5 := s.Insert(2)
	status := "PASS"
	if !(r1 && r2 && !r3 && r4 && r5) {
		status = "FAIL"
	}
	fmt.Printf("%s | Insert(1)=%v Insert(2)=%v Insert(1)=%v Remove(2)=%v Insert(2)=%v\n", status, r1, r2, r3, r4, r5)
}
